assembly/src/library/tests.rs

```diff
@@ -44,23 +46,28 @@ fn masl_locations_serialization() -> Result<(), Report> {
     let modules = vec![foo, bar];
 
     // serialize/deserialize the bundle with locations
-    let namespace = LibraryNamespace::new("test").unwrap();
-    let version = Version::min();
-    let bundle = MaslLibrary::new(namespace, version, modules.iter().cloned(), Vec::new())?;
+    let bundle = Assembler::default().assemble_library(modules.iter().cloned()).unwrap();
```

Comment:
The `masl_locations_serialization()` test currently fails, because the `MastForest` contains 3 procedures, but 4 procedures are exported in the library. This is because `MastForestBuilder::insert_procedure()` will not add 2 procedures with the same root twice. In this test, `test::foo::foo_mul` and `test::bar::bar_mul` have the same root.

This breaks the assumption in `CompiledLibrary.exports` is the same lengths as `CompiledLibrary.mast_forest.num_procedures()`

Reply:
I guess we have 2 options here:

1. Don't allow exports where procedure with the same MAST root is exported under two names. I think there is an argument to be made that if a module exports two identical procedures under different names, this should be an error.
2. Change the `exports` in `CompiledLibrary` to a `BTreeMap` so that we could map multiple procedure names to the same hash.

Reply:
@bitwalker does option 1 add complexity to the compiler?

Reply:
If we have two procedures exported with the same root, that would imply that the "duplicate" is an alias of the other procedure, which might be needed for compatibility purposes (i.e. you want to rename a procedure in your public API, but allow callers to use the old name during the deprecation period).

I think more generally the question is: what do we need to do to properly represent re-exported/aliased procedures in MAST so that the aliases appear like "real" procedures from the perspective of metadata, but resolve to the MAST root of the aliased procedure (which may or may not be in the same MAST forest). My assumption would be that unless we are including the aliased procedure in the same MAST forest, we need to store the MAST digest in the library metadata somewhere, so that requesting that procedure (using the alias), resolves to either the MastNodeId of the aliased procedure, or a MAST digest that can then be looked up in the object store like an external reference.

I think we have to assume that it is perfectly normal for two MAST procedures to have the same MAST root, but originate from different sources, regardless of whether they have the same name or not. On a small scale that may not happen frequently, but the more code there is out there, the more likely it becomes that there exists another function that compiles to the same output (e.g. same function written in two different languages, whose backends produce the same IR for that function).

Reply:
Following up on this: in my commits added to this branch, I've modified `CompiledLibrary` to support re-exported procedures. It was fairly straightforward actually - the only real complications were due to other bugs/issues that were resolved at the same time.

In effect though, what I was proposing is implemented - re-exported procedures appear in the interface of the library, but are stored either as an external MAST root, or the node id corresponding to the aliased procedure in the MAST forest of the library. From an external perspective, re-exported procedures behave just like non-aliased procedures.